### PR TITLE
Allow arbitrary function-like index definitions

### PR DESCRIPTION
**Pull request by geophile**
Generalize grammar to allow for function calls in index DDL, beyond just Z_ORDER_LAT_LON.

### CODE REVIEW

**Comment by nathanlws** [comment]
Looks mostly as expected. This doesn't actually allow anything other GEO_LAT_LON and FULL_TEXT, though, as IndexColumnList.functionType() still goes through Enum.valueOf.

That enum can probably just go away and hang onto the string of the identifier.

A new test with something other than GEO_LAT_LON would be good to add.

File: src/main/javacc/SQLGrammar.jj [hunk 1]
```diff
@@ -8445,58 +8444,59 @@ indexItemList(IndexColumnList columnList) throws StandardException :
 
 void indexItem(IndexColumnList columnList) throws StandardException :
 {
-    String columnName;
+    String firstName = null;
+    String columnName = null;
     boolean asc = true;
-    int spatialColumnCount;
+    int firstSpatialColumnPosition = -1;
```
Comment (nathanlws) [inline src/main/javacc/SQLGrammar.jj]: Minor nit: These aren't spatial specific anymore. Maybe firstFunctionArg, functionColumnCount?

Comment (geophile) [inline src/main/javacc/SQLGrammar.jj]: My plan was to add to the enum as we support additional functions. It's hard to see how that works long-term, assuming FDB eventually supports general-purpose indexes on functions. But that's a long way off, I'm guessing. Do you want me to get rid of the enum now?

I'll address your other comments.

Comment (nathanlws) [inline src/main/javacc/SQLGrammar.jj]: > Do you want me to get rid of the enum now?

I think that makes the most sense. All the enum is adding is restricting what identifier can be used. Allowing any now will mean the parser doesn't need a a change for adding the next few.

**Comment by nathanlws** [comment]
One small nit. Otherwise looks good.

**Comment by geophile** [comment]
The case sensitivity control does not seem to extend to functions declared in indexes. If I remove the upper-casing in IndexColumnList:

testParser[alter-table-add-geo-index-1](com.foundationdb.sql.parser.SQLParserTest)  Time elapsed: 0.071 sec  <<< FAILURE!
org.junit.ComparisonFailure: alter-table-add-geo-index-1 expected:<...e5
		
		methodName: [GEO_LAT_LON
		firstArg: 0
		lastArg: 1
		[0]:		
		com.foundationdb.sql.parser.IndexColumn@7b3300e5
		columnName: c1
		tableName: null
		ascending
		[1]:		
		com.foundationdb.sql.parser.IndexColumn@2e5c649
		columnName: c2
		tableName: null
		ascending
]> but was:<...e5
		
		methodName: [geo_lat_lon
		firstArg: 0
		lastArg: 1
		[0]:		
		com.foundationdb.sql.parser.IndexColumn@7b3300e5
		columnName: c1
		tableName: null
		ascending
		[1]:		
		com.foundationdb.sql.parser.IndexColumn@2e5c649
		columnName: c2
		tableName: null
		ascending]>

**Comment by nathanlws** [comment]
Unless the identifier is quoted the default behavior is to lowercase.

**Comment by geophile** [comment]
Then a number of the {alter,create}-geo-index* tests need quotes, or the function names in the .expected files should be lower-cased. 

jao@mintyzack ~/git/sql-parser $ find src -type f | grep "\.sql" | xargs grep -i geo_lat
src/test/resources/com/foundationdb/sql/parser/alter-table-add-geo-index-2.sql:alter table t1 add Unique index idx2 (geo_lat_lon(c1))
src/test/resources/com/foundationdb/sql/parser/create-table-add-geo-index-2.sql:create table t1 (c1 int, c2 double, c3 bigint, index indx1 (geo_lat_lon(c1)))
src/test/resources/com/foundationdb/sql/parser/create-geo-index-1.sql:create index if not exists idx1 on table1 (geo_lat_lon(c1, c2))
src/test/resources/com/foundationdb/sql/parser/alter-table-add-geo-index-1.sql:alter table t1 add Unique index idx2 (GEO_LAT_LON(c1, c2))
src/test/resources/com/foundationdb/sql/parser/create-geo-index-2.sql:create index if not exists idx1 on table1 (geo_lat_lon(c1))
src/test/resources/com/foundationdb/sql/parser/create-table-add-geo-index-1.sql:create table t1 (c1 int, c2 double, c3 bigint, index indx1 (geo_lat_lon(c1, c2)))
src/test/resources/com/foundationdb/sql/unparser/create-geo-index-1.sql:create index if not exists idx1 on table1 (geo_lat_lon(c1, c2))
src/test/resources/com/foundationdb/sql/unparser/create-geo-index-4.sql:create index if not exists idx1 on table1 (a, b, geo_lat_lon(c1, c2), d, e)
src/test/resources/com/foundationdb/sql/unparser/create-geo-index-2.sql:create index if not exists idx1 on table1 (a, geo_lat_lon(c1, c2))
src/test/resources/com/foundationdb/sql/unparser/create-geo-index-3.sql:create index if not exists idx1 on table1 (geo_lat_lon(c1, c2), b, c)
src/test/resources/com/foundationdb/sql/unparser/create-geo-index-5.sql:create index if not exists idx1 on table1(p.a, p.b, geo_lat_lon(c.c1, c.c2), c.d, c.e) using left join
jao@mintyzack ~/git/sql-parser $ find src -type f | grep "\.expected" | xargs grep -i geo_lat
src/test/resources/com/foundationdb/sql/parser/create-geo-index-2.expected:	methodName: GEO_LAT_LON
src/test/resources/com/foundationdb/sql/parser/create-table-add-geo-index-2.expected:		methodName: GEO_LAT_LON
src/test/resources/com/foundationdb/sql/parser/create-table-add-geo-index-1.expected:		methodName: GEO_LAT_LON
src/test/resources/com/foundationdb/sql/parser/create-geo-index-1.expected:	methodName: GEO_LAT_LON
src/test/resources/com/foundationdb/sql/parser/alter-table-add-geo-index-2.expected:		methodName: GEO_LAT_LON
src/test/resources/com/foundationdb/sql/parser/alter-table-add-geo-index-1.expected:		methodName: GEO_LAT_LON
src/test/resources/com/foundationdb/sql/unparser/create-geo-index-5.expected:CREATE INDEX IF NOT EXISTS idx1 ON table1(p.a, p.b, GEO_LAT_LON(c.c1, c.c2), c.d, c.e) USING LEFT JOIN
src/test/resources/com/foundationdb/sql/unparser/create-geo-index-2.expected:CREATE INDEX IF NOT EXISTS idx1 ON table1(a, GEO_LAT_LON(c1, c2))
src/test/resources/com/foundationdb/sql/unparser/create-geo-index-4.expected:CREATE INDEX IF NOT EXISTS idx1 ON table1(a, b, GEO_LAT_LON(c1, c2), d, e)
src/test/resources/com/foundationdb/sql/unparser/create-geo-index-3.expected:CREATE INDEX IF NOT EXISTS idx1 ON table1(GEO_LAT_LON(c1, c2), b, c)
src/test/resources/com/foundationdb/sql/unparser/create-geo-index-1.expected:CREATE INDEX IF NOT EXISTS idx1 ON table1(GEO_LAT_LON(c1, c2))

**Comment by nathanlws** [comment]
A quick sed on the expected ought to do the trick.

**Comment by nathanlws** [comment]
They are all already changing in this merge, from Z_ -> GEO.

**Comment by nathanlws** [comment]
Thanks.

Approve

**Pull request merged**